### PR TITLE
Feature/4359 create new digital careers pools and candidates

### LINKED ISSUE #4359: Write migration that creates a pool for each Stream/Classification, and assigns existing users to them.
We need to create a new Pool, in production, for each IT Classification / Stream combination. That means 4*8 new pools.

We then need to go through every existing applicant, and assign them to the correct pools, based on the data DCMO entered into their DigitalCareers pool Candidate entities. There should be a mapping between CmoAssets and Job Streams - if a candidate had the right CmoAsset, they can be placed in the matching Stream pools, provided they also match the classification.

To decide if a candidate matches a classification, use the same logic used in the search - expected classification if possible, otherwise do a match on expected salary.

QUESTION: what tools do we want to use to do this? Should the pools be created manually? Should this all be done via a migration? Should it be done in a Seeder (gives more control over when it happens, means we WON'T have these pools in our test environment - is that good or bad)? @petertgiles mentions he made a custom artisan command for skills before, and that might be useful.

## Acceptance Criteria
- [x] The 4*8 Pools are created, as according to Gray's pseudocode
- [ ] We create a new PoolCandidate object for each existing user for each new Pool they can belong to
- [x] PoolCandidates in the existing digital_careers pool all set to expired.

### PR DESCRIPTION
**Pull request by vd1992**
closes #4359 

### Adds three new commands
Create ~32~ 36 new pools with 
`php artisan create:digital_careers_pools`

Create candidates for those pools with
`php artisan create:new_digital_pool_candidates`

Expire candidates for the old pool with 
`php artisan expire:old_digital_pool_candidates`

### Adds new stream "Information and Data Functions"
Added to schema, enums file, and constants; translation pending. 

### Testing
Seeder has been adjusted to test, candidates need assets for example, so migrate + seed fresh before playing around
What exactly to test for, not sure

### CODE REVIEW

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (petertgiles) [inline api/app/Console/Commands/CreateDigitalCareersPools.php:15]: Maybe add a comment on when these commands can be safely deleted in case we forget later.

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (petertgiles) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:64]: You can use the map function to map the $assetCollection in a more functional way, just like in Javascript, and not need to use a foreach loop. https://laravel.com/docs/8.x/collections#method-map

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (petertgiles) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:67]: Nice!  Using a map here is very readable and expressive.

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (petertgiles) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:105]: Big if/else trees like this should probably be a switch instead.  https://www.php.net/manual/en/control-structures.switch.php

File: api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php
Comment (petertgiles) [inline api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php:55]: Nice!  You got the db command parameters going!

**Review by petertgiles — COMMENTED** [review]
Looking good, it seems like you're on the right track!  I'll validate all the logic and test the results when your PR is out of draft but I've added some commentary on the code so far.

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:53]: I commented that it is to changed later here
https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/files#diff-86049c6160cc1408c53fd2a76c59471af238eddcc268ab3d2808ceb2933205ebR60 
but yeah, definitely something easy to mess up, so this is an important comment

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:15]: probably best to make an issue for clearing out commands, I think the comments probably won't stick in memory long

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: introduced sole() here https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/cb29b908941f1c837e3582ece9c18d0be13f36c0

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: reasonable https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/cb29b908941f1c837e3582ece9c18d0be13f36c0

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: php didn't like that

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: mapWithKeys seems to do it 
https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/fc80a132fed5eb8aeaf60d8817dfbee4ad2e0437

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: switch block switched in here https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/fc80a132fed5eb8aeaf60d8817dfbee4ad2e0437

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (petertgiles) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:15]: I've added #4488

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (petertgiles) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: See, the PR went into Review/QA and it still has "admin@test.com" in it.  :sob:

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (petertgiles) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: Oh, I see now.  You're using it to actually share arrays between functions.  Yeah, that would have to be global then.

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: ah, I did sneak this in earlier
https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/4543b6df425c549d8227a47c3be3d80cc96fdd44#diff-86049c6160cc1408c53fd2a76c59471af238eddcc268ab3d2808ceb2933205ebL53-R54
intend to clean that up with feedback

**Review by petertgiles — CHANGES_REQUESTED** [review]
OK, part way through and it's looking really good.  I've noticed a pair of weird issues though.

The first is that it seems like the seeder is assigning all users to a single classification:
image
And that caused me to notice that the second command is putting users out into all the pools even though they should just be in the eight that match that classification:
image
There's definitely mismatches between the pool classification and the classification of the user in it.
image

<details>
  <summary>Some queries I'm testing with:</summary>

```
-- classifications with user count
select classification_id, count(*) 
from classification_user
group by classification_id

-- pool classification vs user classification check
select pc.pool_id, cp.classification_id pool_classification, cu.classification_id user_classification, count(*)
from pools p
left join classification_pool cp on p.id = cp.pool_id
join pool_candidates pc on p.id = pc.pool_id 
left join users u on pc.user_id = u.id 
left join classification_user cu on u.id = cu.user_id 
group by pc.pool_id, cp.classification_id, cu.classification_id
order by pc.pool_id, cp.classification_id, cu.classification_id

-- pools with candidate count
select p.id, p.created_at, p.name->'en', count(*) 
from pool_candidates pc
join pools p on pc.pool_id = p.id
group by p.id, p.created_at, p.name->'en'
order by p.created_at asc
```

</details>

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (tristan-orourke) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:187]:
```suggestion
                [$classificationId, $classificationMinSalary, $classificationMaxSalary] = Classification::where('group', 'ilike', 'IT')
                    ->where('level', $level)
                    ->sole(['id', 'min_salary', 'max_salary']);
```

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:60]: oops 😶 
https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/c3398c2dbd2b2e1fc36a43d9b1739c598f6742af

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: Gray wants them to match their original status so https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/c3398c2dbd2b2e1fc36a43d9b1739c598f6742af

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Great job with the overall logic! I made a number of specific code comments, but it really makes sense overall.

One extra thing - I'd really like a few tests for the CreateDigitalCareersPoolsCandidates command. We're only going to run it once, but we reaaaally want it to run correctly. A phpunit test where we recreate the digital_careers pool, create the new pools, and then create individual pool candidates to check that they're copied to where we think they should be.

File: api/database/seeders/CmoAssetSeeder.php
Comment (vd1992) [reply api/database/seeders/CmoAssetSeeder.php:108]: the seeder changes are temporary, I intend to reverse them later 
those assets weren't on the map so I wasn't sure I needed them for testing

File: api/database/seeders/DatabaseSeeder.php
Comment (vd1992) [reply api/database/seeders/DatabaseSeeder.php:67]: yep, I intend to reverse them later, it was just a quick change for testing purposes so it looks a bit silly
the commits to reverse I named as such https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/3d9c79af455ba40894d12829569ee23d0feb3076 and https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/6e03330d92bfca4e7539343358b35fa339876f1a

File: api/app/Models/Pool.php
Comment (vd1992) [reply api/app/Models/Pool.php:72]: a little tricky in this case as user_id is not optional on pools, you need a real user_id to create a pool

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (petertgiles) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: But none of them have statuses in prod, right?  This is something coming in new for direct-intake?

**Comment by petertgiles** [comment]
> And that caused me to notice that the second command is putting users out into all the pools

:facepalm: That's probably the salary logic, eh?  I'll rereview once the seeder is fixed.

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:53]: it creates an array with the id instead of returning just the id
image

so I think you have to do something like this instead, so not much of a point to picking out just the id
`$ownerId = User::where('sub', 'ilike', 'admin@test.com')->sole(['id'])['id']`

Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: switched sole around a bit to this https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/e7fb1fa68dd7ad770f1663f4a4cbee502866cc88

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: query out of the loop here https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/e7fb1fa68dd7ad770f1663f4a4cbee502866cc88

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: yep https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/e7fb1fa68dd7ad770f1663f4a4cbee502866cc88

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: yep https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/e7fb1fa68dd7ad770f1663f4a4cbee502866cc88

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php]: I have no idea about statuses in the prod environment

**Comment by vd1992** [comment]
big changes in https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/5c04f45196e8b39ca44532b09e20c137bdb3883f
seeder, pass things into functions, etc

**Comment by vd1992** [comment]
php testing added in https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/457ff2029928bd48e4e2f3fef0b256ab022d06e8

File: api/app/Models/Pool.php
Comment (tristan-orourke) [reply api/app/Models/Pool.php:72]: You can do this:
```
$pool = new Pool($fillableData);
$pool->user_id = $user_id;
$pool->save();
```

File: api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php
Comment (petertgiles) [inline api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php:50]: Don't forget to update the `updated_at` timestamp, too.  (UTC, remember)

**Review by petertgiles — CHANGES_REQUESTED** [review]
The new tests are looking cool!

I've done some more querying today and I only noticed one small thing (below).

Three things I'm waiting on:
- The final asset/stream mapping
- The final pool owner
- Confirmation on new pool candidate status
- status for new candidates

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (tristan-orourke) [reply api/app/Console/Commands/CreateDigitalCareersPools.php]: My bad, I thought passing an array into sole would select only those columns for you. I must have misunderstood some documentation or example somewhere.

File: api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php
Comment (vd1992) [reply api/app/Console/Commands/ExpireOldDigitalCareersPoolCandidates.php:50]: got stuff
french translation https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/8f8f229b615ef684a697e7e7dfab66d49b546521
updated_at here https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/467b8ca27f65f3bb0abd7cc21074275890377b73
pool owner here https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/66da2ae68eda458557e1221de45abb023e0cee5b

**Comment by vd1992** [comment]
@tristan-orourke 
move around cmo assets logic https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/bb365f9b4ba15d1f9ea0992efafc02b2a2f2d40c
renaming the test https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/7c285bfb18ff1e1b6dd76e189128f2cd480efefd
remove user from mass assignment https://github.com/GCTC-NTGC/gc-digital-talent/pull/4467/commits/66b83aa9ce3f67a7966a50b721782935ec5c1cb5

**Comment by tristan-orourke** [comment]
Can confirm that the asset-to-stream mapping is complete: https://github.com/GCTC-NTGC/gc-digital-talent/issues/4359#issuecomment-1297446549

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (tristan-orourke) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:150]: This is working, for sure, but I recommend passing in the same `$userModelForCandidate` that we pass to `candidateMatchesClassification`, and getting the asset ids via the Eloquent relation and the very handy [pluck method](https://laravel.com/docs/9.x/collections#method-pluck) on Eloquent collections. I find it more succinct and more legible:
```suggestion
        function candidateMatchesStream(object $userModelForCandidate, string $stream): bool {
            global $assetArrayMap, $assetKeyToStreamMap;

            // get ids of all CmoAssets associated with the user
            $assetsIdArray = $userModelForCandidate->cmoAssets->pluck('id');
```

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (tristan-orourke) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:181]: You're getting the same user twice, here. With the suggestions mentioned above, we can pass $userModelForCandidate to both `candidateMatchesClassification` and `candidateMatchesStream`. If we want to make this slightly more efficient, we can [eager load](https://stackoverflow.com/questions/26005994/laravel-with-method-versus-load-method) the cmoAssets when we get the user initially.
```suggestion
            // We eager load expectedClassifications and cmoAssets because we know we will use them to match Classifications and Streams.
            $userModelForCandidate = User::where('id', $userId)->with('expectedClassifications')->with('cmoAssets')->sole();
```

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (tristan-orourke) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:206]:
```suggestion
                            && candidateMatchesStream($userModelForCandidate, $stream)) {
```

File: api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php
Comment (tristan-orourke) [inline api/app/Console/Commands/CreateDigitalCareersPoolsCandidates.php:91]:
```suggestion
            $candidateClassificationsArray = $userModelForCandidate->expectedClassifications
                                                                    ->pluck('id')
                                                                    ->toArray();
```

**Review by tristan-orourke — APPROVED** [review]
Great job, Vachan! I've left a last few suggestions to improve code fluency, but all my other comments have been addressed. Please wait for Peter's approval as well to merge this in.

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (petertgiles) [inline api/app/Console/Commands/CreateDigitalCareersPools.php:15]: Actually 48 now!  4x8

**Review by petertgiles — APPROVED** [review]
Seems good to me!

File: api/app/Console/Commands/CreateDigitalCareersPools.php
Comment (vd1992) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:15]: huh

Comment (petertgiles) [reply api/app/Console/Commands/CreateDigitalCareersPools.php:15]: Lol, I'm a dummy.  My db client returned 48 rows (36 + the original 12).  Time to take a break...  :coffee:

**Pull request merged**
